Login and Sign up UI

simonrozsival opened the pull request:
## Info

### Summary
I updated the Onboarding storyboard to look like the designs. I changed the presentation style of the Login and Sign Up screens to be page sheets.

There is a hint of a loading state when the login/signup action is being processed, but I haven't implemented that as part of this PR. I'm not sure if it's in the scope of this PR. I added the spinner into the storyboard, but there is no loading part of the login state that could be observed. The loading state should be implemented later.

### Relationships
Closes #602, closes #603, closes #604.

~I haven't added any UI for errors, that should be handled by #601 and in a follow-up request that will solve this also for #603 (that's why it's not closed in this PR). Both the login and sign up screens should most likely be scrollable the same way that #659 defined it, but it's not implemented in this PR.~

- I implemented scrolling
- The error labels are in the UI, but they are not bound to anything yet.

I added a follow-up issue to add the show/hide password buttons: #690

## Repository health

### Pull Request checklist
<!-- MANDATORY: Before submiting your Pull Request, make sure that -->
<!-- all the items below are verified and the boxes are checked -->
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [ ] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

## Reviewing

### Pointers
<!-- MANDATORY: Give pointers to help reviewers validate the changes, give a list of things that should be tested, show before/after screenshots, etc. -->

### Review template
<!-- MANDATORY: This is what reviewers should consider when reviewing your Pull request -->
<!-- Feel free to remove items that UI related if your code does not touch UI -->

<!-- General -->
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones
<!-- UI related -->
- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
⛔ this PR is still a WIP

dshr commented on Modules/Assets/Assets/Storyboards/Base.lproj/Onboarding.storyboard:
```diff
@@ -44,24 +44,100 @@
-                                        <rect key="frame" x="0.0" y="0.0" width="374" height="34"/>
-                                        <fontDescription key="fontDescription" type="system" pointSize="14"/>
-                                        <textInputTraits key="textInputTraits"/>
+                                    <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="Email" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="Jst-Aa-zXL">
```
You can wrap all the inputs and labels into a UIStackView, so we don't have to play around with constraints when showing/hiding the error labels

dshr commented on Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift:
```diff
@@ -101,6 +96,17 @@ public class LoginViewController: UIViewController, Storyboarded {
+        signUpButton.setAttributedTitle(signUpButtonTitle, for: .normal)
+    }
+    
+    @objc private func hideKeyboard() {
```
Re keyboards — we have the`KeyboardObserving` protocol that will help with resizing/repositioning the view's contents.

Odrakir commented:
This one will requiere a lot of changes once #673 is merged. I'm assigning my self as a reviewer to take a look after that's done.

simonrozsival commented:
@dshr I added the scroll view + the code from the reset VC (it might be a good idea to somehow share the code somehow, but I'm not sure how yet..) + I used the stack view for the errors as you suggested.

@Odrakir I changed the navigation and it seems to work the way it's indented to.

dshr commented on Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift:
```diff
@@ -124,3 +148,30 @@ extension LoginViewController: UIAdaptivePresentationControllerDelegate {
     }
 }
+
+extension LoginViewController: KeyboardObserving {
+    public func keyboardWillShow(_ notification: Notification) {
+        guard let keyboardSize = notification.keyboardSize else { return }
+        setScrollViewInsets(keyboardSize.height)
+        scrollView.scrollRectToVisible(CGRect(x: 0,
+                                              y: scrollView.contentSize.height,
+                                              width: 1,
+                                              height: 1),
+                                       animated: true)
+    }
+    
+    public func keyboardWillChangeFrame(_ notification: Notification) {
+        guard let keyboardSize = notification.keyboardSize else { return }
+        setScrollViewInsets(keyboardSize.height)
+    }
+    
+    public func keyboardWillHide(_ notification: Notification) {
+        setScrollViewInsets(0)
+    }
+    
+    private func setScrollViewInsets(_ bottomOffset: CGFloat) {
+        let contentInsets = UIEdgeInsets(top: 0.0, left: 0.0, bottom: bottomOffset, right: 0.0)
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
```
The logic has to be a bit different here, since scrolling to bottom on SE causes the login field to be over the fold:
<img width="432" alt="Screenshot 2020-08-07 at 12 37 45" src="https://user-images.githubusercontent.com/6031512/89632528-04d07900-d8ab-11ea-809e-d6fd7e0fb3f3.png">

simonrozsival replied:
I can't reproduce this anymore on SE2 after the latest changes.

dshr commented on Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift:
```diff
@@ -74,3 +117,30 @@ extension SignupViewController: UIAdaptivePresentationControllerDelegate {
     }
 }
+
+extension SignupViewController: KeyboardObserving {
+    public func keyboardWillShow(_ notification: Notification) {
+        guard let keyboardSize = notification.keyboardSize else { return }
+        setScrollViewInsets(keyboardSize.height)
+        scrollView.scrollRectToVisible(CGRect(x: 0,
+                                              y: scrollView.contentSize.height,
+                                              width: 1,
+                                              height: 1),
+                                       animated: true)
+    }
+    
+    public func keyboardWillChangeFrame(_ notification: Notification) {
+        guard let keyboardSize = notification.keyboardSize else { return }
+        setScrollViewInsets(keyboardSize.height)
+    }
+    
+    public func keyboardWillHide(_ notification: Notification) {
+        setScrollViewInsets(0)
+    }
+    
+    private func setScrollViewInsets(_ bottomOffset: CGFloat) {
+        let contentInsets = UIEdgeInsets(top: 0.0, left: 0.0, bottom: bottomOffset, right: 0.0)
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
+}
```
Something strange is happening here on SE — the scrollview's added insets end up being smaller then needed, covering the login button. This way even though the view looks okay, the users can't scroll down to the login button.
<img width="432" alt="Screenshot 2020-08-07 at 12 38 30" src="https://user-images.githubusercontent.com/6031512/89632696-46f9ba80-d8ab-11ea-9f82-70fd6b814b8f.png">

simonrozsival commented:
I'll double check the font sizes 👍

dshr commented:
> I'll double check the font sizes 👍

Might not be the sizes per se, but font weights as well

dshr commented:
Will do! Meanwhile, I'm working on https://github.com/toggl/ios/issues/601, and thought we could handle it the same way as in https://github.com/toggl/ios/pull/675 where the API layer would transform the `NetworkingError` into an `APIError` which would contain the proper localised description, which you can then pop into `setError` and populate the error label.

dshr commented on Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift:
```diff
@@ -41,6 +45,17 @@ public class SignupViewController: UIViewController, Storyboarded {
         
+        store.select(\.password)
+            .map(isStrongPassword)
+            .map({ strong in strong ? 0.6 : 1.0 })
```
Why not:
```suggestion
            .map({ $0 ? 0.6 : 1.0 })
```

dshr commented on Modules/Assets/Assets/Storyboards/Base.lproj/Onboarding.storyboard:
```diff
@@ -122,117 +335,340 @@
+                                                    </textField>
+                                                    <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="Include 8 or more characters, with at least one uppercase letter, lowercase letter, and a number." textAlignment="natural" lineBreakMode="tailTruncation" numberOfLines="0" baselineAdjustment="alignBaselines" adjus ...
+                                                        <rect key="frame" x="0.0" y="186" width="311" height="50.333333333333343"/>
+                                                        <fontDescription key="fontDescription" type="system" weight="medium" pointSize="14"/>
```
This should be regular, not medium. Still, I think Sketch has rendered the letter-spacing wrong, but there's not much we can do about it.
```suggestion
                                                        <fontDescription key="fontDescription" type="system" pointSize="14"/>
```
